k2/csrc/array_inl.h

```diff
@@ -49,6 +49,37 @@ void Array1<T>::CopyFrom(const Array1<T> &src) {
   MemoryCopy(static_cast<void *>(dst_data), static_cast<const void *>(src_data),
              Dim() * ElementSize(), kind, Context().get());
 }
+template <typename T>
+template <typename S>
+Array1<S> Array1<T>::AsType() {
+  Array1<S> ans(Context(), Dim());
+  S *ans_data = ans.Data();
+  const T *this_data = Data();
+  auto lambda_set_values = [=] __host__ __device__(int32_t i) {
+    ans_data[i] = static_cast<S>(this_data[i]);
+  };
+  Eval(Context(), Dim(), lambda_set_values);
+  return ans;
+}
+
+// if S is same with T, just return *this. But in C++, we cannot explicitly
+// specialize a class member template unless its enclosing class templates are
+// also explicitly specialized. Thus here we specialized Array1 as well.
+template <>
```

Comment:
This is unnecessary, you can just do
```
  if (std::is_same<S, T>) return *this;
```
at the top.

Reply:
mm. we can't do this，`if` is runtime code, not compiling time. If T is double and S is float, You are trying to return double array for a function returns float array. noted the code  `if (true) return *this` and below codes always get into the generated code

Reply:
you could try: return *(reinterpret_cast<Array1<S> >(this));

Reply:
Also, you should be able to do the template specialization as 
`template <typename S> void Array1<S>::AsType<S>();`
but I think the if thing is maybe clearer, and should be optimized to the same thing.

Reply:
OK, use cast here. Noted we cannot specialize, that's exactly what I commented before:
https://github.com/k2-fsa/k2/blob/d5e8626c51d7f21d55cc3c0e1516efdf832c6a99/k2/csrc/array_inl.h#L65-L67

Reply:
Actually, you can use SIFNA

```cpp
  template <typename S>
  Array1<typename std::enable_if<!std::is_same<S, T>::value, S>::type>
  AsType() const {
    //  S != T
    // .... 
  }
  template <typename S>
  Array1<typename std::enable_if<std::is_same<S, T>::value, S>::type> AsType() const {
    // S == T
    return *this;
  }
```
No cast is needed.

Reply:
The cast results in the following warning:

```
k2/k2/csrc/array_inl.h(56): warning: dynamic initialization in unreachable code
          detected during:
            instantiation of "k2::Array1<S> k2::Array1<T>::AsType<S>() [with T=double, S=double]"
```

Reply:
Can you paste the code your are compiling with? It's strange that I got no such warning with T=double and S=double.

BTW, seems I missed your last comments. I'm OK with using SFINA as long as Dan is fine with this.

Reply:
I am using the default settings and compiling against the master branch.

Maybe the warning depends on the compiler we are using.

Reply:
Here is my local version: `gcc 5.5.0` `nvcc 10.1`. Anyway, I'm OK with either SFINA or just supressing the warning (as we know it's fine to supress it). It would be good if @danpovey can give his comment.

Reply:
I'm OK with either.

Reply:
Updated to use SFINA in PR https://github.com/k2-fsa/k2/pull/286/files